File: Python/oomph-data-ws/uf_309_lib/ufs/int_beta/unionfind_103.py
class UnionFind:
    """
    Represents a disjoint-set data structure, using union by size and path compression.

    time complexity: O(α(n))
    """

    def __init__(self, n: int):
        self.parents = [-1] * n
        self.sizes = [1] * n

    def find(self, x: int):
        if self.parents[x] == -1:
            return x
        else:
            self.parents[x] = self.find(self.parents[x])
            return self.parents[x]

    def are_same(self, x: int, y: int):
        return self.find(x) == self.find(y)

    def get_group_size(self, x: int):
        return self.sizes[self.find(x)]

    def union(self, x: int, y: int):
        x = self.find(x)
        y = self.find(y)
        if x == y:
            return False

        if self.sizes[x] < self.sizes[y]:
            x, y = y, x
        self.parents[y] = x
        self.sizes[x] += self.sizes[y]
        return True

File: Java/Oomph.Data/UF09Lib/UFs/IntBeta/UnionFind.103.cs
// path compression, union by size
// O(α(n))

public static class UnionFind
{
	int[] parents, sizes;

	public UnionFind(int n)
	{
		parents = new int[n];
		Arrays.fill(parents, -1);
		sizes = new int[n];
		Arrays.fill(sizes, 1);
	}

	public int Find(int x)
	{
		return parents[x] == -1 ? x : (parents[x] = Find(parents[x]));
	}
	public boolean AreSame(int x, int y)
	{
		return Find(x) == Find(y);
	}
	public int GetGroupSize(int x)
	{
		return sizes[Find(x)];
	}

	public boolean Union(int x, int y)
	{
		if ((x = Find(x)) == (y = Find(y))) return false;

		if (sizes[x] < sizes[y])
		{
			int t = x;
			x = y;
			y = t;
		}
		parents[y] = x;
		sizes[x] += sizes[y];
		return true;
	}
}
